### ISSUE TITLE
turn off verbose for clustering and my questions about clustering

### ISSUE THREAD

**Issue by hainm** [Amber-MD/cpptraj#62]
hi,

this is ``pytraj`` got when calling kmeans in ``cpptraj``

```python
In [4]: from pytraj.cluster import kmeans

In [5]: kmeans(traj, n_clusters=3)
#Clustering: 3 clusters 10 frames
#Cluster 0 has average-distance-to-centroid 5.643003
#Cluster 1 has average-distance-to-centroid 5.645873
#Cluster 2 has average-distance-to-centroid 4.141970
#DBI: 1.298084
#pSF: 3.236341
#Algorithm: Kmeans nclusters 3 maxit 100
#Representative frames: 9 5 1

Out[5]: array([2, 2, 1, 1, 1, 1, 0, 0, 0, 0], dtype=int32)
```

I although ``pytraj`` turn off cpptraj's verbose but still get the stdout. 

* Can we turn it off (if yes, I can make PR, or it's ok if it's quick for you to do).

* Does you intend to include clustering info (average-centroid-distance, refp frame number, ...) in Dataset? (currently they are printed to stdout).

* I have an idea that try to delay writing cluster trajs to disk (only do when calling ``Print``). What do you think about performance. 

* I would like to construct ``pytraj`` workflow like this
   * loading ``traj``
   * perform clustering and getting frame indcies for each cluster, getting rep frame number
   * after playing with other stuff, calling ``write_cluster`` for specific clusters (for example, top 5)

What do you think?

Examples code: mostly I just need to write to pdb files to view in VMD. So for rep frames, I can write all to a single file like this.

```python
rep_frame_indices = get_from_clustering...
traj[ref_frame_indices].save('reps.pdb', mode='model')
```

**Comment by hainm** [comment]
My motivation is simple: to expose cpptraj more to ``sklearn``. For example. to expose to

[MiniBatchKMeans](http://scikit-learn.org/stable/modules/generated/sklearn.cluster.MiniBatchKMeans.html)

**Comment by drroe** [comment]
> I although pytraj turn off cpptraj's verbose but still get the stdout.

This is because the clustering info output uses a version of `CpptrajFile` that defaults to stdout if no file name is specified. This is done both to be consistent with previous ptraj behavior, and also because it stinks when you do a large clustering run and forget to specify an output file and the bulk of your results are lost to the ether.

> Can we turn it off (if yes, I can make PR, or it's ok if it's quick for you to do).

Since it doesn't use mprintf() it is not as simple to turn off and I don't want to change the behavior. I would say either have pytraj write the cluster info file or I (or you) will have to add an option to clustering specifically to suppress output. The former is the easiest and least intrusive option in the short term, but the latter could be implemented without too much pain if necessary. 

> Does you intend to include clustering info (average-centroid-distance, refp frame number, ...) in Dataset? (currently they are printed to stdout).

Maybe after the dataset_metadata branch is merged. That should happen relatively soon. 

> I have an idea that try to delay writing cluster trajs to disk (only do when calling Print). What do you think about performance.

Not sure what you mean here. Currently writing cluster trajectories happens after clustering only if any cluster traj output options were specified. How much more do you want to delay? You can also write out cluster trajectories completely separately from clustering using the cluster num v time dataset and the `outtraj` command - is that more like what you mean?

**Comment by hainm** [comment]
>  I (or you) will have to add an option to clustering specifically to suppress output. 

I can work on this and I prefer this way.

> Maybe after the dataset_metadata branch is merged. That should happen relatively soon.

cool

> You can also write out cluster trajectories completely separately from clustering using the cluster num v time dataset and the outtraj command - is that more like what you mean?

thanks. I will look into this.

**Comment by hainm** [comment]
@drroe 

I've just tried Action_Outtraj. Do you think I should use this as work-horse for writing trajectory in ``pytraj``? It's high level API so it's likely you won't change its much?

**Comment by hainm** [comment]
``outtraj``, ``trajout``. :D

**Comment by drroe** [comment]
> Do you think I should use this as work-horse for writing trajectory in pytraj?

It depends on if you need the `maxmin` functionality of `outtraj`; if not it's more direct to use Trajout_Single (one less level of function calls). Unfortunately there's nothing that says the Action API won't change. In fact since the first version of cpptraj it has changed at least twice. Although you are right that it is less likely to change...

> outtraj, trajout. :D

They do almost the same thing which is why the names are so similar, but `outtraj` respects the order in which it was specified in the Action list while `trajout` always outputs at the end. So with
```
trajout something
rms first
```
the output coords are changed by `rms`, but with
```
outtraj somthing
rms first
```
they are not.

**Comment by hainm** [comment]
thanks.

On Mon, Sep 7, 2015 at 11:23 AM, Daniel R. Roe <notifications@github.com>
wrote:

> Do you think I should use this as work-horse for writing trajectory in
> pytraj?
>
> It depends on if you need the maxmin functionality of outtraj; if not
> it's more direct to use Trajout_Single (one less level of function calls).
> Unfortunately there's nothing that says the Action API won't change. In
> fact since the first version of cpptraj it has changed at least twice.
> Although you are right that it is less likely to change...
>
> outtraj, trajout. :D
>
> They do almost the same thing which is why the names are so similar, but
> outtraj respects the order in which it was specified in the Action list
> while trajout always outputs at the end. So with
>
> trajout something
> rms first
>
> the output coords are changed by rms, but with
>
> outtraj somthing
> rms first
>
> they are not.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Amber-MD/cpptraj/issues/62#issuecomment-138324591>.
>



-- 
Hai Nguyen

**Comment by hainm** [comment]
closed by #182
